From 15f630adc4325f9cafb762e15a051c2ec8bd21e2 Mon Sep 17 00:00:00 2001
From: Sutou Kouhei <kou@clear-code.com>
Date: Mon, 21 Dec 2020 04:30:44 +0900
Subject: [PATCH] Use RAKE_EXTENSION_TASK_NO_NATIVE env var as the default
 value

---
 lib/rake/extensiontask.rb | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/rake/extensiontask.rb b/lib/rake/extensiontask.rb
index b0ab88e..49d59da 100644
--- a/lib/rake/extensiontask.rb
+++ b/lib/rake/extensiontask.rb
@@ -23,7 +23,7 @@ module Rake
       @cross_compile = false
       @cross_config_options = []
       @cross_compiling = nil
-      @no_native = false
+      @no_native = (ENV["RAKE_EXTENSION_TASK_NO_NATIVE"] == "true")
       @config_includes = []
       @ruby_versions_per_platform = {}
       @make = nil
-- 
2.27.0

